ckanext/cfpb_extrafields/plugin.py

```diff
@@ -76,9 +76,16 @@ def _redirect_to_edit_on_change(self, resource, field):
     def _delete_and_rebuild_datadict(self, resource):
         import json
         if 'datadict' in resource and 'id' in resource:
-            record = resource['datadict']
-            resource.pop('datadict')
-            json_record = json.loads(record)
+            record = resource.pop('datadict')
+
+            try:
+                json_record = json.loads(record)
+            except ValueError as e:
+                # Invalid JSON, so don't remove old data
+                error_message = "Error saving data dictionary: {0}.  Data was: {1}".format(e, record)
+                flash_error(error_message)
```

Comment:
I had this happen when adding a data dictionary to a new dataset, no matter what I put into the fields.  I think this is because the two stage dataset creation puts the data dictionary update logic into a weird state.  I will look into it more tomorrow to see if there's a better fix.